### ISSUE TITLE
Concatenate CNN (trained using Keras flow_from_directory and generators) and LSTM (time-series data)

### ISSUE THREAD

**Issue by NKM999** [keras-team/keras#15179]

**Comment by NKM999** [comment]
Hello,

I am trying to develop a model using multimodal data. My one set of data is images that is trained using Keras flow_from_directory and generators) and another set of data is time-series that is trained using LSTM and CONV1D. How to concatenate the two models to train them as a single model (model with two inputs and single output)? Any example or guidance will assist in its practical implementation.

Thanks and Regards

**Comment by patel-rupin2000** [comment]
Hey I tried to merge two models I have merge LSTM and CNN(in which VGG16 has been used) and it works
  join = Concatenate([lstm_model, cnn_model])
  hidden = Dense(1, activation = 'sigmoid')
  new_model = Sequential()
  new_model.add(join)
  new_model.add(hidden)
  new_model.compile(optimizer="",loss="",metrices=[""])

  history = new_model .fit(X_train, y_train, epochs=50, batch_size=64) 
  import matplotlib.pyplot as plt
  plt.plot(history.history['accuracy'],label="train_accuracy")
  plt.plot(history.history['val_accuracy'],label="test_accuracy")
  plt.legend()

Or their is another method you can use any one of the model as base model  just like VGG16 is used for CNN
base_model =  VGG16(input_shape = (224, 224, 3), include_top = False, weights = 'imagenet')
for layer in base_model.layers:
    layer.trainable = False
x = Flatten()(base_model.output)
x = Dense(1024, activation='relu')(x)
x = Dropout(0.5)(x)
x = Dense(3, activation='softmax')(x)
model = Model(base_model.input, x)
model.compile(optimizer = "adam", loss = 'categorical_crossentropy',metrics = ["accuracy"])

**Comment by pankaj-cipher** [comment]
We try to concatenate things up when they are of same property or shape. 
So, if we try to concatenate `flow_from_directry()` to `LSTM` and `Conv1D` it would not make sense as one is for data augmentation and other is not .
So, basically we kind of try to do this thing of concatenation for monitoring some gestures or movements over video which basically involve temporal and spatial domains here temporal feature extraction is done by ` LSTM` , `Conv1D` and spatial features extraction is done by `Conv2D` .
So, the solution is do the data augmentation before training and store images somewhere , now for every image you would be having time series data , but you have to generate for augmented images .
After that make a parallel model as it is after all the calculation just concatenate the features together which could be done easily .
and then pass it to dense to predict .
I can do a pull request which can make it more simpler for you to understand!

**Comment by NKM999** [comment]
> Hey I tried to merge two models I have merge LSTM and CNN(in which VGG16 has been used) and it works
> join = Concatenate([lstm_model, cnn_model])
> hidden = Dense(1, activation = 'sigmoid')
> new_model = Sequential()
> new_model.add(join)
> new_model.add(hidden)
> new_model.compile(optimizer="",loss="",metrices=[""])
> 
> history = new_model .fit(X_train, y_train, epochs=50, batch_size=64)
> import matplotlib.pyplot as plt
> plt.plot(history.history['accuracy'],label="train_accuracy")
> plt.plot(history.history['val_accuracy'],label="test_accuracy")
> plt.legend()
> 
> Or their is another method you can use any one of the model as base model just like VGG16 is used for CNN
> base_model = VGG16(input_shape = (224, 224, 3), include_top = False, weights = 'imagenet')
> for layer in base_model.layers:
> layer.trainable = False
> x = Flatten()(base_model.output)
> x = Dense(1024, activation='relu')(x)
> x = Dropout(0.5)(x)
> x = Dense(3, activation='softmax')(x)
> model = Model(base_model.input, x)
> model.compile(optimizer = "adam", loss = 'categorical_crossentropy',metrics = ["accuracy"])

Thanks for the reference code.

**Comment by NKM999** [comment]
> We try to concatenate things up when they are of same property or shape.
> So, if we try to concatenate `flow_from_directry()` to `LSTM` and `Conv1D` it would not make sense as one is for data augmentation and other is not .
> So, basically we kind of try to do this thing of concatenation for monitoring some gestures or movements over video which basically involve temporal and spatial domains here temporal feature extraction is done by ` LSTM` , `Conv1D` and spatial features extraction is done by `Conv2D` .
> So, the solution is do the data augmentation before training and store images somewhere , now for every image you would be having time series data , but you have to generate for augmented images .
> After that make a parallel model as it is after all the calculation just concatenate the features together which could be done easily .
> and then pass it to dense to predict .
> I can do a pull request which can make it more simpler for you to understand!

Thanks for the valuable suggestions.
I am searching if there is any library or solution available to concatenate CNN and LSTM. Also, I want to get the benefit of the Keras image data generator that provides data augmentation and improves performance considerably. By doing manual data augmentation, I will not be able to get diversity over each iteration. Basically, I want to improve images (CNN) performance by adding LSTM in parallel. However, without the Keras image data generator, substantial improvement may not be possible.

**Comment by alankritmishra** [comment]
I have been trying to figure out how can I merge two input data( one image dataset another time-series dataset) and tried to use CNN and LSTM. The number of samples in the image dataset is less than that of time-series data. But is there a way to use concrete both models for classification? I have labels for 40 classes in both datasets.

**Comment by tilakrayal** [comment]
@NKM999,
Could you please try below code where we are trying to merge two Deep learning models

 ```
model1 = Sequential()
    #input 
    model1.add(Dense(32, input_shape=(NUM_FEAT1,1)))
    model1.add(Activation("elu"))
    model1.add(Dropout(0.5))
    model1.add(Dense(16))
    model1.add(Activation("elu"))
    model1.add(Dropout(0.25))
    model1.add(Flatten())

    model2 = Sequential()
    #input 
    model2.add(Dense(32, input_shape=(NUM_FEAT1,1)))
    model2.add(Activation("elu"))
    model2.add(Dropout(0.5))
    model2.add(Dense(16))
    model2.add(Activation("elu"))
    model2.add(Dropout(0.25))
    model2.add(Flatten())

merged = Concatenate()([model1.output,model2.output])
z = Dense(128, activation="relu")(merged)
z = Dropout(0.25)(z)
z = Dense(1024, activation="relu")(z)
z = Dense(1, activation="sigmoid")(z)

model = Model(inputs=[model1.input, model2.input], outputs=z)

model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])


        model.fit([x_train[train_index][:,:66], x_train[train_index][:,66:132], y_train[train_index], batch_size=100, epochs=100, verbose=2)
```

We could able to feed 2 different types of data to the model like images in the first model and Textual Data in the second model accordingly.

Also please take a look at this [issue](https://github.com/keras-team/keras/issues/401) with a similar request for the reference. Thank you!
